pkg/bucketaccessrequest/bucketaccessrequest.go

```diff
@@ -0,0 +1,141 @@
+	glog.V(1).Infof("update called for bucket %v", old)
+	return nil
+}
+
+func (b *bucketAccessRequestListener) Delete(ctx context.Context, obj *v1alpha1.BucketAccessRequest) error {
+	glog.V(1).Infof("delete called for bucket %v", obj)
+	return nil
+}
+
+// provisionBucketAccess  attempts to provision a BucketAccess for the given bucketAccessRequest.
+// Returns nil error only when the bucketaccess was provisioned. An error is return if we cannot create bucket access.
+// A normal error is returned when  bucket acess  was not provisioned and provisioning should be retried (requeue the bucketAccessRequest),
+// or a special error  errBucketAccessAlreadyExists, errInvalidBucketAccessClass is returned when provisioning was impossible and
+// no further attempts to provision should be tried.
+func (b *bucketAccessRequestListener) provisionBucketAccess(ctx context.Context, bucketAccessRequest *v1alpha1.BucketAccessRequest) error {
+	bucketAccessClassName := bucketAccessRequest.Spec.BucketAccessClassName
+
+	bucketaccess := b.FindBucketAccess(ctx, bucketAccessRequest)
+	if bucketaccess != nil {
+		// bucketaccess has been already provisioned, nothing to do.
+		return util.ErrBucketAccessAlreadyExists
+	}
+
+	bucketAccessClass, err := b.bucketClient.ObjectstorageV1alpha1().BucketAccessClasses().Get(ctx, bucketAccessClassName, metav1.GetOptions{})
+	if bucketAccessClass == nil {
+		// bucket has been already provisioned, nothing to do.
+		return util.ErrInvalidBucketAccessClass
+	}
+
+	bucketRequest, err := b.bucketClient.ObjectstorageV1alpha1().BucketRequests(bucketAccessRequest.Namespace).Get(ctx, bucketAccessRequest.Spec.BucketRequestName, metav1.GetOptions{})
+	if bucketRequest == nil {
+		// bucket request does not exist, we have to reject this provision.
+		return util.ErrInvalidBucketRequest
+	}
+	if err != nil {
+		return err
+	}
+
+	if bucketRequest.Spec.BucketInstanceName == "" {
+		return util.ErrWaitForBucketProvisioning
+	}
+
+	bucketaccess = &v1alpha1.BucketAccess{}
+	bucketaccess.Name = util.GetUUID()
+
+	bucketaccess.Spec.BucketInstanceName = bucketRequest.Spec.BucketInstanceName
+	bucketaccess.Spec.BucketAccessRequest = bucketAccessRequest.Name
+	bucketaccess.Spec.ServiceAccount = bucketAccessRequest.Spec.ServiceAccountName
+	// bucketaccess.Spec.MintedSecretName - set by the driver
+	bucketaccess.Spec.PolicyActionsConfigMapData, err = util.ReadConfigData(b.kubeClient, bucketAccessClass.PolicyActionsConfigMap)
+	if err != nil {
+		return err
+	}
+	//  bucketaccess.Spec.Principal - set by the driver
+
+	bucketaccess.Spec.Provisioner = bucketAccessClass.Provisioner
+	bucketaccess.Spec.Parameters = util.CopySS(bucketAccessClass.Parameters)
+
+	bucketaccess, err = b.bucketClient.ObjectstorageV1alpha1().BucketAccesses().Create(context.Background(), bucketaccess, metav1.CreateOptions{})
+	if err == nil || apierrs.IsAlreadyExists(err) {
```

Comment:
If the err is "already exists" then there is probably more work to do for idempotency.  We need to do some checks to make sure the existing BA is what we expect.

Reply:
Fixed.